### ISSUE TITLE
Incorrect return type for core next() function

### ISSUE THREAD

**Issue by jnvsor** [vimeo/psalm#738]
I'm getting false positive MoreSpecificReturnType/LessSpecificReturnStatement issues from `next()` but not from `reset()`

Both `next()` and `reset()` are supposed to return an array element or false on failure. Since the parameter in question is typehinted as `array[]` the result of `reset($trace)` must be `false|array`.

The same should apply to `next($trace)` but psalm is identifying that as 'mixed' causing the false positive. Here's a demonstration:

```php
<?php

/**
 * @param array[] $trace Backtrace
 *
 * @return array{callee:false|array, caller:false|array} Call info
 */
function test(array $trace)
{
    return [
        'callee' => reset($trace),
        'caller' => next($trace),
    ];
}
```

```
INFO: LessSpecificReturnStatement - 10:5 - The type 'array{callee:false|array<mixed, mixed>, caller:mixed}' is more general than the declared return type 'array{callee:array<mixed, mixed>|false, caller:array<mixed, mixed>|false}' for test

INFO: MoreSpecificReturnType - 6:12 - The declared return type 'array{callee:array<mixed, mixed>|false, caller:array<mixed, mixed>|false}' for test is more specific than the inferred return type 'array{callee:false|array<mixed, mixed>, caller:mixed}'
```
